File: DSA/DSA_450/07_binary_search_trees/016_Count_Nodes_In_Range.cpp
/*
Problem: Count BST Nodes That Lie in a Given Range
URL: https://practice.geeksforgeeks.org/problems/count-bst-nodes-that-lie-in-a-given-range/1

Problem Statement:
Count BST nodes that lie in a given range [low, high].

Sample Input/Output:
Input: root = [10,5,50,1,null,40,100], low = 5, high = 45
Output: 3
Explanation: Nodes in range are 5, 10, 40.
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int data;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : data(x), left(NULL), right(NULL) {}
};

TreeNode* Insert_BST(TreeNode* root, int key) {
    if (root == NULL) return new TreeNode(key);
    if (key < root->data) root->left = Insert_BST(root->left, key);
    else root->right = Insert_BST(root->right, key);
    return root;
}

void Print_Inorder(TreeNode* root) {
    if (root == NULL) return;
    Print_Inorder(root->left);
    cout << root->data << " ";
    Print_Inorder(root->right);
}

class Solution {
public:
    int Count_Nodes_Pruned(TreeNode* root, int low, int high) {
        /*
        Pruned traversal approach
        Time Complexity: O(h + k) where k is nodes in range
        Space Complexity: O(h)
        */
        if (root == NULL) return 0;
        if (root->data >= low && root->data <= high) {
            return 1 + Count_Nodes_Pruned(root->left, low, high) + 
                   Count_Nodes_Pruned(root->right, low, high);
        } else if (root->data < low) {
            return Count_Nodes_Pruned(root->right, low, high);
        } else {
            return Count_Nodes_Pruned(root->left, low, high);
        }
    }

    int Count_Nodes_Full(TreeNode* root, int low, int high) {
        /*
        Full traversal approach
        Time Complexity: O(n)
        Space Complexity: O(h)
        */
        if (root == NULL) return 0;
        int count = 0;
        if (root->data >= low && root->data <= high) count++;
        count += Count_Nodes_Full(root->left, low, high);
        count += Count_Nodes_Full(root->right, low, high);
        return count;
    }
};

void Test_Count_Nodes_In_Range() {
    Solution solution;
    TreeNode* root = NULL;
    root = Insert_BST(root, 10);
    root = Insert_BST(root, 5);
    root = Insert_BST(root, 50);
    root = Insert_BST(root, 1);
    root = Insert_BST(root, 40);
    root = Insert_BST(root, 100);
    int low = 5, high = 45;
    cout << "Count Nodes (Pruned): " << solution.Count_Nodes_Pruned(root, low, high) << endl;
    cout << "Count Nodes (Full): " << solution.Count_Nodes_Full(root, low, high) << endl;
}

int main() {
    Test_Count_Nodes_In_Range();
    return 0;
}

File: DSA/DSA_450/Python/07_binary_search_trees/016_Count_Nodes_In_Range.py
"""
Problem: Count BST Nodes That Lie in a Given Range
URL: https://practice.geeksforgeeks.org/problems/count-bst-nodes-that-lie-in-a-given-range/1

Problem Statement:
Count BST nodes that lie in a given range [low, high].

Sample Input/Output:
Input: root = [10,5,50,1,null,40,100], low = 5, high = 45
Output: 3
Explanation: Nodes in range are 5, 10, 40.
"""


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    @staticmethod
    def Build_BST(keys):
        root = None
        for key in keys:
            root = TreeNode.Insert_BST(root, key)
        return root

    @staticmethod
    def Insert_BST(root, key):
        if root is None:
            return TreeNode(key)
        if key < root.val:
            root.left = TreeNode.Insert_BST(root.left, key)
        else:
            root.right = TreeNode.Insert_BST(root.right, key)
        return root

    @staticmethod
    def Print_Inorder(root):
        if root is None:
            return
        TreeNode.Print_Inorder(root.left)
        print(root.val, end=" ")
        TreeNode.Print_Inorder(root.right)


class Solution:
    def Count_Nodes_Pruned(self, root, low, high):
        """
        Pruned traversal approach
        Time Complexity: O(h + k) where k is nodes in range
        Space Complexity: O(h)
        """
        if root is None:
            return 0
        if root.val >= low and root.val <= high:
            return (1 + self.Count_Nodes_Pruned(root.left, low, high) +
                    self.Count_Nodes_Pruned(root.right, low, high))
        elif root.val < low:
            return self.Count_Nodes_Pruned(root.right, low, high)
        else:
            return self.Count_Nodes_Pruned(root.left, low, high)

    def Count_Nodes_Full(self, root, low, high):
        """
        Full traversal approach
        Time Complexity: O(n)
        Space Complexity: O(h)
        """
        if root is None:
            return 0
        count = 0
        if root.val >= low and root.val <= high:
            count += 1
        count += self.Count_Nodes_Full(root.left, low, high)
        count += self.Count_Nodes_Full(root.right, low, high)
        return count


def Test_Count_Nodes_In_Range():
    solution = Solution()
    root = None
    root = TreeNode.Insert_BST(root, 10)
    root = TreeNode.Insert_BST(root, 5)
    root = TreeNode.Insert_BST(root, 50)
    root = TreeNode.Insert_BST(root, 1)
    root = TreeNode.Insert_BST(root, 40)
    root = TreeNode.Insert_BST(root, 100)
    low, high = 5, 45
    print("Count Nodes (Pruned):", solution.Count_Nodes_Pruned(root, low, high))
    print("Count Nodes (Full):", solution.Count_Nodes_Full(root, low, high))


if __name__ == "__main__":
    Test_Count_Nodes_In_Range()
